guides/developer/crd.md

```diff
@@ -0,0 +1,211 @@
+These extension points live alongside the config-based support for CRDs described above.
+
+This work has allowed us to support the following features within Spinnaker:
+
+ - Custom models for representing CRDs as `spinnakerKinds`.
+ - Deploying CRDs with custom Spinnaker artifact types.
+ - Custom Kubernetes API versions.
+ - Custom Spinnaker naming strategies.
+ - Per-account, custom Spinnaker UIs that can run alongside the existing Kubernetes UIs.
+
+This guide exists for developers that want to duplicate this functionality for their CRDs.
+It also exists an explanation of certain code paths within Spinnaker which include hooks with no corresponding open-source implementations.
+
+Developers that want to implement these features will have to build their own layered version
+of [Clouddriver](https://github.com/spinnaker/clouddriver) -
+see Adam Jorden's [blog post](https://blog.spinnaker.io/scaling-spinnaker-at-netflix-custom-features-and-packaging-e78536d38040) - and should be familiar with the Kubernetes V2 provider and writing code for Clouddriver.
+
+## Custom Handlers
+
+The central extension point is the `KubernetesHandler` class. A subclass of `KubernetesHandler` - e.g., `KubernetesReplicaSetHandler` - defines the
+relationship between Spinnaker and your Kubernetes kind.
+
+For example, if you wanted to build a Spinnaker integration for your CRD of kind `MyCRDKind`, you would start with
+the following handler:
+
+```
+@Component
+public class MyCRDHandler extends KubernetesHandler {
+
+  public MyCRDHandler() {
+    // Hook point for registering a custom `ArtifactReplacer`
+    // for your CRD. During a deploy operation,
+    // if an artifact of the type specified in the replacer is present,
+    // the artifact will be inserted into the manifest using the
+    // strategy described in the replacer.
+  }
+
+  @Override
+  public KubernetesKind kind() {
+    return "MyCRDKind";
+  }
+
+  @Override
+  public boolean versioned() {
+    // If the CRD resource should be versioned - i.e., assigned a sequence
+    // v001, v002, etc.
+    return false;
+  }
+
+  @Override
+  public SpinnakerKind spinnakerKind() {
+    // The Spinnaker kind that the resource will be represented as in Spinnaker's API and UI.
+    return SpinnakerKind.SERVER_GROUPS;
+  }
+
+  @Override
+  public Status status(KubernetesManifest manifest) {
+    // Includes logic for determining whether your CRD manifest is stable.
+    // A deploy manifest operation, for example, will block until this
+    // method returns a stable status.
+  }
+
+  @Override
+  public Class<? extends KubernetesV2CachingAgent> cachingAgentClass() {
+    // Caching agent class for your CRD.
+    // See, e.g., `KubernetesReplicaSetCachingAgent`.
+    return MyCRDCachingAgent.class;
+  }
+}
+```
+
+## Custom Spinnaker Resource Models
+
+Developers may want to change how their CRD is represented in Spinnaker's API. By default, a CRD of `spinnakerKind` `serverGroups` will
```

Comment:
Safe to assume everyone reading this doc (with the intention of using it) is a developer? So change to "You might want to change how your CRD is represented..."?